opencti-platform/opencti-graphql/src/schema/schema-attributes.ts

```diff
@@ -218,3 +217,82 @@ export const isObjectFlatAttribute = (k: string): boolean => {
 export const isMultipleAttribute = (entityType: string, k: string): boolean => (
   k.startsWith(RULE_PREFIX) || schemaAttributesDefinition.isMultipleAttribute(entityType, k)
 );
+
+// -- utility functions independent of attribute registration --
+// (inner mappings are not registered like first-level attribute)
+
+export const isMandatoryAttributeMapping = (schemaDef: AttributeDefinition) => schemaDef.mandatoryType === 'external' || schemaDef.mandatoryType === 'internal';
+
+export const isNonFlatObjectAttributeMapping = (schemaDef: AttributeDefinition) : schemaDef is ComplexAttributeWithMappings => { // handy typeguard
+  return schemaDef.type === 'object' && schemaDef.format !== 'flat';
+};
+
+/**
+ * Validates that the given input conforms to the constraints in the corresponding schema definition.
+ * Recursively checks non-flat objects mappings.
+ * @param input an input object candidate to indexing in elastic
+ * @param schemaDef AttributeDefinition for the given input data
+ */
+const validateInputAgainstSchema = (input: any, schemaDef: AttributeDefinition) => {
+  const isMandatory = isMandatoryAttributeMapping(schemaDef);
+  if (isMandatory && R.isNil(input)) {
+    throw FunctionalError(`Validation against schema failed on attribute [${schemaDef.name}]: this mandatory field cannot be nil`, { value: input });
+  }
+
+  if (isNonFlatObjectAttributeMapping(schemaDef)) {
+    // check 'multiple' constraint
+    if (isMandatory) {
+      if (schemaDef.multiple && !Array.isArray(input)) {
+        throw FunctionalError(`Validation against schema failed on attribute [${schemaDef.name}]: value must be an array`, { value: input });
+      }
+      if (!schemaDef.multiple && (Array.isArray(input) || !R.is(Object, input))) {
+        throw FunctionalError(`Validation against schema failed on attribute [${schemaDef.name}]: value must be an object`, { value: input });
+      }
+    }
+    if (!isMandatory && R.isNil(input)) {
+      return; // nothing to check (happens on 'remove' operation for instance
+    }
```

Comment:
The last test you wrote failed because the function checks multiple constraint only if the attribute is mandatory. But, thanks to your test, I think it is a mistake because in the case the attribute is an object, not mandatory, and you give an input which is not an object then we don't throw.

The suggestion is to first check is not mandatory and no value then return, otherwise check the type of the input no matter what the mandatory type is.

```suggestion
    if (!isMandatory && R.isNil(input)) {
      return; // nothing to check (happens on 'remove' operation for instance
    }
    // check 'multiple' constraint
    if (schemaDef.multiple && !Array.isArray(input)) {
      throw FunctionalError(`Validation against schema failed on attribute [${schemaDef.name}]: value must be an array`, { value: input });
    }
    if (!schemaDef.multiple && (Array.isArray(input) || !R.is(Object, input))) {
      throw FunctionalError(`Validation against schema failed on attribute [${schemaDef.name}]: value must be an object`, { value: input });
    }
```